File: spring-boot-sample/src/main/java/com/example/demo/controller/ProductController.java
package com.example.demo.controller;

import com.example.demo.model.Product;
import com.example.demo.service.ProductService;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;

@RestController
public class ProductController {

    private final ProductService svc;

    public ProductController(ProductService svc) {
        this.svc = svc;
    }

    @GetMapping("/api/products")
    public List<Product> getProducts() {
        return svc.getAll();
    }

    @GetMapping("/api/products-delayed")
    public List<Product> getProductsDelayed() {
        return svc.getAllDelayed();
    }

    @GetMapping("/api/products/{id}")
    public Product getProduct(@PathVariable long id) {
        return svc.getById(id);
    }

    @GetMapping("/actuator/health")
    public String health() {
        return "UP";
    }
}

File: dotnet8-sample/Controllers/ProductController.cs
using Microsoft.AspNetCore.Mvc;
using dotnet_sample.Services;
using dotnet_sample.Models;
using System.Diagnostics;

namespace dotnet_sample.Controllers;

[ApiController]
[Route("api/[controller]")]
public class ProductsController : ControllerBase
{
    private readonly ProductService _svc;
    private static readonly ActivitySource ActivitySource = new("dotnet-sample.ProductController");

    public ProductsController(ProductService svc)
    {
        _svc = svc;
    }

    [HttpGet]
    public ActionResult<IEnumerable<Product>> Get()
    {
        using var activity = ActivitySource.StartActivity("get-all-products");
        var products = _svc.GetAll();
        activity?.SetTag("product.count", products.Count());
        return Ok(products);
    }

    [HttpGet("delayed")]
    public async Task<ActionResult<IEnumerable<Product>>> GetDelayed()
    {
        using var activity = ActivitySource.StartActivity("get-all-products-delayed");
        activity?.SetTag("operation.type", "slow");
        var products = await _svc.GetAllDelayed();
        activity?.SetTag("products.returned", products.Count());
        return Ok(products);
    }

    [HttpGet("{id:long}")]
    public ActionResult<Product?> GetById(long id)
    {
        using var activity = ActivitySource.StartActivity("get-product-by-id");
        activity?.SetTag("product.id", id);
        var p = _svc.GetById(id);
        if (p == null) 
        {
            activity?.SetTag("product.found", false);
            return NotFound();
        }
        activity?.SetTag("product.found", true);
        activity?.SetTag("product.name", p.Name);
        return Ok(p);
    }

    [HttpGet("/actuator/health")]
    public ActionResult<string> Health() => Ok("UP");
}
